examples/dialogflow-webhook-example/main.py

```diff
@@ -0,0 +1,445 @@
+    """Return the account balance."""
+    flattener = kwargs['flattener']
+    messages = Messages()
+
+    user_id = self._get_user_id(flattener)
+
+    # Get firestore client and extract the user data base on user_id provided.
+    db = get_firestore_client()
+    client_ref = db.collection(u'users').where(u'user_id', u'==', user_id)
+
+    users = client_ref.stream()
+    for user in users:
+      # Example:
+      # user_dict: {
+      #   'user_id': 123456,
+      #   'first_name': 'Pedro',
+      #   'last_name': 'Perez',
+      #   'accounts': {
+      #       'checking': {
+      #         'transactions': [
+      #             {'amount': 50, 'type': 'deposit'},
+      #             {'type': 'withdraw', 'amount': '-10'}
+      #           ],
+      #           'balance': 150
+      #        },
+      #        'saving': {
+      #           'transactions': [
+      #               {'amount': 20, 'type': 'deposit'},
+      #               {'type': 'deposit', 'amount': 90}
+      #            ],
+      #            'balance': 110
+      #         }
+      #     }
+      # }
+      user_dict = user.to_dict()
+      return json.dumps({
+          'fulfillmentText':
+              'From webhook: Here are your account balances. Checking account is {0}  Saving account is {1}. What else can I do for you?'
+              .format(user_dict['accounts']['checking']['balance'],
+                      user_dict['accounts']['saving']['balance'])
+      })
+
+    return messages.user_id_not_found(user_id)
+
+  def account_transactions_handler(self, **kwargs):
+    """Returns all the transactions (deposits and withdraws)."""
+    flattener = kwargs['flattener']
+    messages = Messages()
+
+    # Get the user_id from the context in the request.
+    user_id = self._get_user_id(flattener)
+
+    # Get the firestore client and extract the user data base on user_id
+    # provided.
+    db = get_firestore_client()
+    client_ref = db.collection(u'users').where(u'user_id', u'==', user_id)
+    users = client_ref.stream()
+    for user in users:
+      # Example:
+      # user_dict: {
+      #   'user_id': 123456,
+      #   'first_name': 'Pedro',
+      #   'last_name': 'Perez',
+      #   'accounts': {
+      #       'checking': {
+      #         'transactions': [
+      #             {'amount': 50, 'type': 'deposit'},
+      #             {'type': 'withdraw', 'amount': '-10'}
+      #           ],
+      #           'balance': 150
+      #        },
+      #        'saving': {
+      #           'transactions': [
+      #               {'amount': 20, 'type': 'deposit'},
+      #               {'type': 'deposit', 'amount': 90}
+      #            ],
+      #            'balance': 110
+      #         }
+      #     }
+      # }
+      user_dict = user.to_dict()
+      transactions = user_dict['accounts']['checking']['transactions']
+      message = 'From webhook: Here are all the transactions that I found:\n'
+      checking_transactions = ''
```

Comment:
L300-319 is redundant, verbose, memory inefficient (allocating a list of two elements in every iteration just to put a comma between the latest item and the rest of the items aggregated in the string), slow (string.format is slower than f-string) and leads to some quite confusing variable naming / reuse. Instead this should be DRY-d up / greatly simplified using list concatenation with `+` operator and list comprehension. This means we we only allocate 1 list. This is much more readable (and faster)
```python3
      all_transactions = ','.join([
          '{} {}'.format(transaction['type'], transaction['amount'])
          for transaction in user_dict['accounts']['checking']['transactions'] +
          user_dict['accounts']['saving']['transactions']
      ])
      return json.dumps({
          'fulfillmentText':
              ' '.join([
                  message, all_transactions, 'What else can I do for you?'])})
```
and the following benchmark shows it's significantly faster :)
```
old:  9.91780741896946
new:  6.51355233398499
```

```python3 
import timeit
user_dict = {
    'accounts': {
        'checking': {
            'transactions': [{
                'type': 'foo',
                'amount': i
            } for i in range(10)]
        },
        'saving': {
            'transactions': [{
                'type': 'bar',
                'amount': i 
            } for i in range(10)]
        }
    }
}

def old():
    checking_transactions = ''
    saving_transactions = ''
    transactions = user_dict['accounts']['checking']['transactions']
    for transaction in transactions:
      checking_transactions = ','.join([
          '{} {}'.format(transaction['type'], transaction['amount']),
          checking_transactions
      ])
    transactions = user_dict['accounts']['saving']['transactions']
    for transaction in transactions:
      saving_transactions = ','.join([
          '{} {}'.format(transaction['type'], transaction['amount']),
          saving_transactions
      ])

def new():
    all_transactions = ','.join([
        f"{transaction['type']} {transaction['amount']}"
        for transaction in user_dict['accounts']['checking']['transactions'] +
        user_dict['accounts']['saving']['transactions']
    ])
print('old: ', timeit.timeit(old))
print('new: ', timeit.timeit(new))
```